makeflow/src/makeflow_hook.h

```diff
@@ -0,0 +1,322 @@
+	 * @param jx the job specification as encoded like a wrapper (see json).
+	 * @return 1 if successfully modified, 0 if not.
+	 */
+	int (*batch_submit) ( struct batch_queue *, struct jx *);
+
+
+	/* Hook after node is collected, but prior to qualifying node success.
+	 * 
+	 * This hook occurs after node is retrieved from batch queue. This
+	 * allows hooks to perform on outputted files and see exit status.
+	 *
+	 * @param dag_node The dag_node that was collected.
+	 * @param batch_job_info The info struct passed by batch queue.
+	 * @return 1 if successful, 0 if not.
+	 */
+	int (*node_end)      (struct dag_node *, struct batch_job_info *);
+
+	/* Hook if node was successful.
+	 * 
+	 * @param dag_node The dag_node that was successful.
+	 * @param batch_job_info The info struct passed by batch queue.
+	 * @return 1 if successful, 0 if not.
+	 */
+	int (*node_success)  (struct dag_node *, struct batch_job_info *);
+
+	/* Hook if node failed.
+	 * 
+	 * @param dag_node The dag_node that failed.
+	 * @param batch_job_info The info struct passed by batch queue.
+	 * @return 1 if successful, 0 if not.
+	 */
+	int (*node_fail)     (struct dag_node *, struct batch_job_info *);
+
+	/* Hook if node aborted.
+	 * 
+	 * @param dag_node The dag_node that was aborted.
+	 * @param batch_job_info The info struct passed by batch queue.
+	 * @return 1 if successful, 0 if not.
+	 */
+	int (*node_abort)    (struct dag_node *, struct batch_job_info *);
+
+	/* Hook when file is created.
+	 * 
+	 * Allows modifications when file is created.
+	 *
+	 * @param dag_file The dag_file that was initialized.
+	 * @return 1 is successful, 0 if not.
+	 */
+	int (*file_create)   (struct dag_file *);
+
+	/* Hook when file is expected, prior to node submission.
+	 *
+	 * @param dag_file The dag_file that is expected.
+	 * @return 1 is successful, 0 if not.
+	 */
+	int (*file_expect)   (struct dag_file *);
+
+	/* Hook when file is registered as existing.
+	 *
+	 * @param dag_file The dag_file that exists.
+	 * @return 1 is successful, 0 if not.
+	 */
+	int (*file_exist)    (struct dag_file *);
+
+	/* Hook when file is registered as complete.
+	 *
+	 * Complete means that the file still exists, 
+	 * but is no longer used. The next step is to 
+	 * clean it.
+	 *
+	 * @param dag_file The dag_file that completed.
+	 * @return 1 is successful, 0 if not.
+	 */
+	int (*file_complete) (struct dag_file *);
+
+	/* Hook when file is about to be clean.
+	 *
+	 * This is used for archiving files are storing files prior
+	 * to removal.
+	 *
+	 * @param dag_file The dag_file that is to be cleaned.
+	 * @return 1 is successful, 0 if not.
+	 */
+	int (*file_clean)    (struct dag_file *);
+
+	/* Hook when file has been deleted.
+	 *
+	 * @param dag_file The dag_file that is to be cleaned.
+	 * @return 1 is successful, 0 if not.
+	 */
+	int (*file_deleted)  (struct dag_file *);
+	
+};
+
+typedef enum {
+    MAKEFLOW_HOOK_SUCCESS = 0,
+    MAKEFLOW_HOOK_FAILURE
+} makeflow_hook_return;
```

Comment:
Should the hook functions return `makeflow_hook_return` rather than `int`? (And just being picky: if so, I think something like `makeflow_hook_result` would read better)